code/datums/action.dm

```diff
@@ -0,0 +1,357 @@
+#define AB_CHECK_RESTRAINED 1
+#define AB_CHECK_STUNNED 2
+#define AB_CHECK_LYING 4
+#define AB_CHECK_CONSCIOUS 8
+
+
+/datum/action
+	var/name = "Generic Action"
+	var/obj/target = null
+	var/check_flags = 0
+	var/processing = 0
+	var/obj/screen/movable/action_button/button = null
+	var/button_icon = 'icons/mob/actions.dmi'
+	var/button_icon_state = "default"
+	var/background_icon_state = "bg_default"
+	var/mob/living/owner
+
+/datum/action/New(Target)
+	target = Target
+	button = new
+	button.linked_action = src
+	button.name = name
+
+/datum/action/Destroy()
+	if(owner)
+		Remove(owner)
+	target = null
+	qdel(button)
+	button = null
+	return ..()
+
+/datum/action/proc/Grant(mob/living/L)
+	if(owner)
+		if(owner == L)
+			return
+		Remove(owner)
+	owner = L
+	L.actions += src
+	if(L.client)
+		L.client.screen += button
+	L.update_action_buttons()
+
+/datum/action/proc/Remove(mob/living/L)
+	if(L.client)
+		L.client.screen -= button
+	button.moved = FALSE //so the button appears in its normal position when given to another owner.
+	L.actions -= src
+	L.update_action_buttons()
+	owner = null
```

Comment:
We no longer delete and recreate the action's button every time it's removed and granted again.